### PR TITLE
Allow to specify a cpu pressure usage value

### PR DESCRIPTION
**Pull request by taihuynh167**
## What does this PR do?

- [x] Adds new functionality
- [ ] Alters existing functionality
- [ ] Fixes a bug
- [ ] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:
- CPU pressure disruption currently does not take any parameter, and always uses as much CPU as possible. 
- This PR enables users to specify a specific value, which will only fill a certain amount of the CPU. 
- This amount would be specified as a percentage (of CPU to fill) or as an integer (number of cores to fill).

## Code Quality Checklist

- [ ] The documentation is up to date.
   - To be updated
- [x] My code is sufficiently commented and passes continuous integration checks.
- [x] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [ ] I leveraged continuous integration testing
    - [x] by depending on existing `unit` tests or `end-to-end` tests.
    - [ ] by adding new `unit` tests or `end-to-end` tests.
       - I plan to separately unit test the `StresserManager`. The cpu_pressure.go now delegates most of the stuff to `StresserManager`.
- [ ] I manually tested the following steps:
    - [x] locally.
```
{"level":"info","ts":1666618997157.501,"caller":"cgroup/cgroup.go:59","message":"reading from cgroup file","disruptionName":"cpu-pressure","disruptionNamespace":"chaos-demo","targetName":"demo-curl-79bf7b74c9-4m54c","targetNodeName":"colima","path":"/mnt/cgroup/cpuset//kubepods/burstable/pod31634024-44d6-4da8-b549-b4c8cb624f59/529ff60ee9eec61fe3d5edaca93c6ac8b84c3dc08ef214666da62dc59e77ae51/cpuset.cpus"}
{"level":"info","ts":1666618997157.7095,"caller":"stress/cpuset_manager.go:87","message":"retrieved coresToBeStressed after applying userRequestCount 50%: 0-1","disruptionName":"cpu-pressure","disruptionNamespace":"chaos-demo","targetName":"demo-curl-79bf7b74c9-4m54c","targetNodeName":"colima"}
{"level":"info","ts":1666618997158.0562,"caller":"stress/cpuset_manager.go:49","message":"changed GOMAXPROCS value from 4 to 2","disruptionName":"cpu-pressure","disruptionNamespace":"chaos-demo","targetName":"demo-curl-79bf7b74c9-4m54c","targetNodeName":"colima"}
{"level":"info","ts":1666618997158.147,"caller":"injector/cpu_pressure.go:124","message":"joining target CPU cgroup","disruptionName":"cpu-pressure","disruptionNamespace":"chaos-demo","targetName":"demo-curl-79bf7b74c9-4m54c","targetNodeName":"colima","core":3,"pid":31299}
```
    - [ ] as a canary deployment to a cluster.

### CODE REVIEW

File: examples/cpu_pressure.yaml [hunk 1]
```diff
@@ -13,4 +13,6 @@ spec:
   selector:
     app: demo-curl
   count: 1
-  cpuPressure: {} # consume as much CPU as possible to generate load
+  cpuPressure: {
+    count: 50%
+  }
```
Comment (luphaz) [inline examples/cpu_pressure.yaml:18]:
```suggestion
  cpuPressure:
    count: 50%
```

File: stress/cpuset_manager.go [hunk 2]
```diff
@@ -78,7 +81,28 @@ func (manager *cpuStressserManager) TrackInjectorCores(cgroup cgroup.Manager) (c
...
 	}
 
-	manager.setCoresToBeStressed(manager.coresToBeStressed.Union(cores))
+	coresToBeStressedCount, _ := manager.userRequestedCores(cores.Size(), userRequestCount)
+	coresToBeStressed := cpuset.NewCPUSet(cores.ToSlice()[:coresToBeStressedCount]...)
+
+	manager.log.Infof("retrieved coresToBeStressed after applying userRequestCount %s: %s", userRequestCount.StrVal, coresToBeStressed)
+	manager.setCoresToBeStressed(manager.coresToBeStressed.Union(coresToBeStressed))
 
 	return cores, nil
```
Comment (luphaz) [inline stress/cpuset_manager.go:90]: Shouldn't this now be `coresToBeStressed`?
```suggestion
	return coresToBeStressed, nil
```

File: stress/cpuset_manager.go [hunk 3]
```diff
@@ -78,7 +81,28 @@ func (manager *cpuStressserManager) TrackInjectorCores(cgroup cgroup.Manager) (c
...
 }
+
+func (manager *cpuStressserManager) userRequestedCores(cores int, userRequestedCores *intstr.IntOrString) (int, error) {
+	targetCores := 0
+	userRequestedCount, isPercent, err := v1beta1.GetIntOrPercentValueSafely(userRequestedCores)
```
Comment (luphaz) [inline stress/cpuset_manager.go:95]: Are we able to use the method from `apimachinery` somehow:
* https://github.com/kubernetes/apimachinery/blob/master/pkg/util/intstr/intstr.go#L150

Seems they are doing what you are?

Should we also get rid of our own method if possible?

Also the method I suggested seems to already exists in our helpers:
* https://github.com/DataDog/chaos-controller/blob/main/controllers/helpers.go#L43

Maybe using this directly would be good enough if we do not intend to use `apimachinery` or can't

File: injector/cpu_pressure_test.go [hunk 4]
```diff
@@ -51,7 +51,7 @@ var _ = Describe("Failure", func() {
 		manager.On("ThreadID").Return(666)
 
 		stresserManager = &stress.StresserManagerMock{}
-		stresserManager.On("TrackInjectorCores", mock.Anything).Return(cpuset.NewCPUSet(0, 1), nil)
+		stresserManager.On("TrackInjectorCores", mock.Anything, mock.Anything).Return(cpuset.NewCPUSet(0, 1), nil)
```
Comment (luphaz) [inline injector/cpu_pressure_test.go:54]: Adding unit tests to highlight/validate the new behaviour would be a plus.

**Review by luphaz — COMMENTED** [review]
The implementation suggest for every container we'll stress half the assigned CPU

It seems OK from my POV, I would still appreciate more reviews from people having broadest knowledge here

Maybe updating the doc:
* https://github.com/DataDog/chaos-controller/blob/main/docs/cpu_pressure.md

Or FAQ to explain the behaviour could help users of this feature

File: stress/cpuset_manager.go (hunk 2 above)
Comment (taihuynh167) [reply stress/cpuset_manager.go:90]: yes, good spot!

File: stress/cpuset_manager.go (hunk 3 above)
Comment (taihuynh167) [reply stress/cpuset_manager.go:95]: nice, I didn't discover `helpers.go`, I will see to use this now. Thanks!

File: injector/cpu_pressure_test.go (hunk 4 above)
Comment (taihuynh167) [reply injector/cpu_pressure_test.go:54]: i'm setting up new tests for `stresserManager` 🙈 i'm unable to include them here (confession: i still should) as it took me some time today with weird errors. I sent this PR in an early state to get implementation feedback 🙇

**Review by aymericDD — APPROVED** [review]
LGTM :+1:

File: injector/cpu_pressure_test.go [hunk 5]
```diff
@@ -98,27 +100,46 @@ var _ = Describe("Failure", func() {
...
+			BeforeEach(func() {
+				userRequestCount := intstr.FromString("50%")
+				spec = v1beta1.CPUPressureSpec{
+					Count: &userRequestCount,
+				}
+				stresserManager.On("TrackInjectorCores", mock.Anything, &userRequestCount).Return(cpuset.NewCPUSet(0, 1), nil)
+			})
+
+			It("should call stresserManager track cores and get new core to apply pressure", func() {
+				// left empty for AfterEach
```
Comment (luphaz) [inline injector/cpu_pressure_test.go:141]: Assertions of this `It` are only checked by `AssertExpectations` correct? I would rephrase the comment, at first I did not directly grasped why it was left empty 😅  
```suggestion
				// left empty as AfterEach 'AssertExpectations' check all this tests expectations
```

**Review by luphaz — APPROVED** [review]
